### PR TITLE
feat!: replace calico with cilium

### PR DESCRIPTION
**Pull request by onedr0p**
Important changes:

_**kube-vip** is now a core cluster component and not managed by Flux. Read thru the comments on [this](https://github.com/onedr0p/flux-cluster-template/pull/740) PR to discover why this change was needed._

- Deploy and manage Cilium with [k3s HelmChart CR](https://docs.k3s.io/helm) (this might be taken over and managed by flux in the future)
- Deploy and manage kube-vip as a [static pod](https://unofficial-kubernetes.readthedocs.io/en/latest/concepts/cluster-administration/static-pod/) instead of a daemonset. kube-vip is no longer managed by Flux.

Closes: https://github.com/onedr0p/flux-cluster-template/issues/739

### CODE REVIEW

**Comment by ModestTG** [comment]
Here is the error I get when I ge5t to the `TASK [xanmanning.k3s : Check that the control plane is available to accept connections]` step in the `task ansible:install`

```
May 24 03:13:48 k8s-1 k3s[3942]: time="2023-05-24T03:13:48Z" level=fatal msg="starting kubernetes: preparing server: failed to get CA certs: Get \"https://10.0.0.61:6443/cacerts\": dial tcp 10.0.0.61:6443: connect: no route to host"
```
This is occuring on both of the non-primary control nodes. The control node starts up fine.

10.0.0.61 is my kube-vip address. It's not currently occupied by any other host and there are no ACL's blocking traffic (nodes are 10.0.0.[50.55]). Not sure if this is an error with my setup or if this is cilium causing this problem. Let me know if you want me to do any additional steps

**Comment by mrueg** [comment]
I think bootstrapping might be easier by running `helm template | kubectl apply -f` via ansible to avoid chicken and egg problems.

**Comment by onedr0p** [comment]
@mrueg That is how the Talos folks are doing it. I was thinking about it but getting it in ansible might be a challenge.  Alternatively we could use the [cilium-cli](https://docs.cilium.io/en/v1.13/gettingstarted/k8s-install-default/#install-cilium) too but this would also need to be tasks in ansible as well. 🤔 Either option is really crappy here if you also want flux to take it over. I noticed kubespray uses the cli tool to install cilium so maybe there's something to be learned there and its lifecycle is managed outside flux :/

**Comment by mrueg** [comment]
@onedr0p cilium-cli does have an experimental helm mode that could be useful: https://github.com/cilium/cilium-cli#experimental-helm-installation-mode

**Comment by onedr0p** [comment]
That looks interesting and it's coming with the next release, also they are pushing for OCI charts too. Since there's no rush on this I'll need to think about this more but I'm definitely open to ideas.

**Comment by onedr0p** [comment]
I just noticed there is a `spec.bootstrap` flag that can be set on the HelmChart CR.

https://docs.k3s.io/helm

I've commited this and hopefully can be tested soon.

**Comment by onedr0p** [comment]
Adding `spec.bootstrap` definitely helped moved the cilium pods into progression but I was only able to test that far on my Mac M2 + Parallels before running into issue with kube-vip not working over L2.

If anyone is able to help test, that would be great.

**Comment by alex-matthews** [comment]
With 3 control-plane nodes and 1 worker node (all running Ubuntu 22.04.2 LTS), I get the following error upon running `task ansible:install`:

```
TASK [xanmanning.k3s : Check that the control plane is available to accept connections] **********************************************************************************************************************************
skipping: [k8s-3]
ok: [k8s-0]
fatal: [k8s-2]: FAILED! => changed=false 
  elapsed: 300
  msg: Timeout when waiting for 127.0.0.1:6443
fatal: [k8s-1]: FAILED! => changed=false 
  elapsed: 300
  msg: Timeout when waiting for 127.0.0.1:6443

NO MORE HOSTS LEFT *******************************************************************************************************************************************************************************************************

PLAY RECAP ***************************************************************************************************************************************************************************************************************
k8s-0                      : ok=74   changed=15   unreachable=0    failed=0    skipped=55   rescued=0    ignored=0   
k8s-1                      : ok=55   changed=12   unreachable=0    failed=1    skipped=54   rescued=0    ignored=0   
k8s-2                      : ok=55   changed=12   unreachable=0    failed=1    skipped=54   rescued=0    ignored=0   
k8s-3                      : ok=52   changed=12   unreachable=0    failed=0    skipped=57   rescued=0    ignored=0   

task: Failed to run task "ansible:install": exit status 2
```

With 1 control-plane node and 1 worker node, `task ansible:install` completes without throwing errors, but I can't reach the API server at the kube-vip IP address.

After editing the kubeconfig file to point to the IP address of the control-plane node, `kubectl get nodes` returns the following:

```
NAME    STATUS     ROLES                       AGE     VERSION
k8s-0   NotReady   control-plane,etcd,master   2m24s   v1.27.1+k3s1
```

And `kubectl get pods -A` returns the following:

```
NAMESPACE     NAME                              READY   STATUS                  RESTARTS       AGE
kube-system   cilium-operator-5bd7dc76b-9vbrs   0/1     Pending                 0              40m
kube-system   cilium-operator-5bd7dc76b-gknht   0/1     CrashLoopBackOff        12 (80s ago)   40m
kube-system   cilium-r8z94                      0/1     Init:CrashLoopBackOff   10 (97s ago)   40m
kube-system   coredns-77ccd57875-l7q29          0/1     Pending                 0              41m
kube-system   helm-install-cilium-x77zn         0/1     Completed               0              41m
kube-system   hubble-relay-76b7cfb479-wkbzq     0/1     Pending                 0              40m
kube-system   hubble-ui-5d7c6975c7-bpfgn        0/2     Pending                 0              40m
kube-system   kube-vip-jqgz6                    1/1     Running                 0              41m
```

Because I can't even ping the VIP, I guess kube-vip isn't working properly (although the pod seems fine). This is why the Cilium pods are failing. Here is a snippet of what `kubectl describe pod -n kube-system cilium-operator-5bd7dc76b-gknht` returns:

```
level=info msg="Started gops server" address="127.0.0.1:9891" subsys=gops
level=info msg="Start hook executed" duration="896.642µs" function="gops.registerGopsHooks.func1 (cell.go:44)" subsys=hive
level=info msg="Establishing connection to apiserver" host="https://192.168.1.200:6443" subsys=k8s-client
level=info msg="Establishing connection to apiserver" host="https://192.168.1.200:6443" subsys=k8s-client
level=info msg="Establishing connection to apiserver" host="https://192.168.1.200:6443" subsys=k8s-client
level=info msg="Establishing connection to apiserver" host="https://192.168.1.200:6443" subsys=k8s-client
level=info msg="Establishing connection to apiserver" host="https://192.168.1.200:6443" subsys=k8s-client
level=info msg="Establishing connection to apiserver" host="https://192.168.1.200:6443" subsys=k8s-client
level=info msg="Establishing connection to apiserver" host="https://192.168.1.200:6443" subsys=k8s-client
level=info msg="Establishing connection to apiserver" host="https://192.168.1.200:6443" subsys=k8s-client
level=info msg="Establishing connection to apiserver" host="https://192.168.1.200:6443" subsys=k8s-client
level=error msg="Unable to contact k8s api-server" error="Get \"https://192.168.1.200:6443/api/v1/namespaces/kube-system\": dial tcp 192.168.1.200:6443: connect: no route to host" ipAddr="https://192.168.1.200:6443" subsys=k8s-client
level=error msg="Start hook failed" error="Get \"https://192.168.1.200:6443/api/v1/namespaces/kube-system\": dial tcp 192.168.1.200:6443: connect: no route to host" function="client.(*compositeClientset).onStart" subsys=hive
level=info msg="Stopped gops server" address="127.0.0.1:9891" subsys=gops
level=info msg="Stop hook executed" duration="793.535µs" function="gops.registerGopsHooks.func2 (cell.go:51)" subsys=hive
level=fatal msg="failed to start: Get \"https://192.168.1.200:6443/api/v1/namespaces/kube-system\": dial tcp 192.168.1.200:6443: connect: no route to host" subsys=cilium-operator-generic
```

Changing the value of k8sServiceHost in the custom HelmChart template (to point to the IP address of the control-plane node) and re-running `task ansible:install` fixes it:

```
NAMESPACE     NAME                               READY   STATUS      RESTARTS   AGE
kube-system   cilium-mx6ts                       1/1     Running     0          2m2s
kube-system   cilium-operator-6947cc77f9-jjrmg   1/1     Running     0          2m2s
kube-system   cilium-operator-6947cc77f9-wlxnc   1/1     Running     0          2m2s
kube-system   cilium-t92xt                       1/1     Running     0          41s
kube-system   coredns-77ccd57875-mz64p           1/1     Running     0          3m9s
kube-system   helm-install-cilium-mb8pk          0/1     Completed   0          3m10s
kube-system   hubble-relay-76b7cfb479-9gf8b      1/1     Running     0          2m2s
kube-system   hubble-ui-5d7c6975c7-zxh7g         2/2     Running     0          2m2s
kube-system   kube-vip-gwwkw                     1/1     Running     0          3m10s
```

Hope this helps.

**Comment by onedr0p** [comment]
@alex-matthews thanks for the details!

It looks like you might have hit the same thing I did on my basic parallels attempts. Kube vip was spitting out errors on trying to connect to coredns `10.101.0.1 dial tcp timeout`.  So coredns is in a pending state due to no cilium not working due to kube vip not working due to coredns not working 😰

The fix you have is great but I'm not sure how automated it can be with Ansible.

Are you able to test and reboot your nodes after you got it all working? I'm curious if everything is still healthy too.

**Comment by onedr0p** [comment]
I'm seeing one blog post about moving kube vip to a static pod instead of deploying it the normal way. I'll give this a shot however it means flux will not be able to manage the lifecycle of updates for the kube-vip (which is _fine_ I suppose but not ideal) 

https://dischord.org/2022/01/04/k3s-kube-vip-cilium-egress/

**Comment by onedr0p** [comment]
I've pushed changes to deploy `kube-vip` as a static pod and removed flux from managing it.

**Comment by szinn** [comment]
I would suggest going the other way. To me, bringing up a cluster is two distinct steps. First, bringing up a collection of nodes with kids - nothing fancy, no additional components. The CNI would be required here. The second step is bringing up the “application” layer. This is what the flux bootstrap (using the term loosely, not necessary the bootstrap command).

So, with that in mind, installing cilium would happen in the first step, installing kube-vip would be in the second step. After step 1, the cluster is healthy without kube-vip, which just adds some HA robustness.

**Comment by onedr0p** [comment]
I agree but it since cilium cannot function without a LB it's hard to follow that logic here.

**Comment by onedr0p** [comment]
I am going to go ahead and merge this, thanks all for the comments and support <3
